File: IniLoader.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace FLASC
{
    public sealed class IniSection
    {
        public string Name { get; }
        public List<KeyValuePair<string, string>> Entries { get; } = new List<KeyValuePair<string, string>>();

        public IniSection(string name)
        {
            Name = name;
        }

        public string? Get(string key)
        {
            foreach (var kv in Entries)
                if (string.Equals(kv.Key, key, StringComparison.OrdinalIgnoreCase))
                    return kv.Value;
            return null;
        }
    }

    public static class IniLoader
    {
        /// <summary>
        /// Parses the whole .ini file into an ordered list of sections, each with an
        /// ordered list of key/value pairs (order matters: for gameAutomobile/gameHeli/
        /// gameBoat, field order in the .ini is the same as the field's byte offset order
        /// in the binary struct).
        /// </summary>
        public static List<IniSection> Load(string path)
        {
            var sections = new List<IniSection>();
            IniSection? current = null;

            foreach (var rawLine in File.ReadLines(path))
            {
                string line = rawLine.Trim();
                if (line.Length == 0)
                    continue;

                if (line.StartsWith("[", StringComparison.Ordinal) && line.EndsWith("]", StringComparison.Ordinal))
                {
                    current = new IniSection(line.Substring(1, line.Length - 2).Trim());
                    sections.Add(current);
                    continue;
                }

                int eq = line.IndexOf('=');
                if (eq < 0 || current == null)
                    continue;

                string key = line.Substring(0, eq).Trim();
                string value = line.Substring(eq + 1).Trim();
                current.Entries.Add(new KeyValuePair<string, string>(key, value));
            }

            return sections;
        }
    }
}

File: IniLoader.cpp
#include "IniLoader.h"

#include <algorithm>
#include <cctype>
#include <fstream>
#include <sstream>
#include <stdexcept>

namespace flasc {

namespace {

std::string Trim(const std::string& s) {
    size_t start = 0;
    size_t end = s.size();
    while (start < end && std::isspace(static_cast<unsigned char>(s[start])))
        ++start;
    while (end > start && std::isspace(static_cast<unsigned char>(s[end - 1])))
        --end;
    return s.substr(start, end - start);
}

bool EqualsIgnoreCase(const std::string& a, const std::string& b) {
    if (a.size() != b.size())
        return false;
    for (size_t i = 0; i < a.size(); ++i) {
        if (std::tolower(static_cast<unsigned char>(a[i])) !=
            std::tolower(static_cast<unsigned char>(b[i])))
            return false;
    }
    return true;
}

} // namespace

std::optional<std::string> IniSection::Get(const std::string& key) const {
    for (const auto& kv : entries) {
        if (EqualsIgnoreCase(kv.first, key))
            return kv.second;
    }
    return std::nullopt;
}

namespace IniLoader {

std::vector<IniSection> Load(const std::string& path) {
    std::ifstream file(path, std::ios::in | std::ios::binary);
    if (!file.is_open())
        throw std::runtime_error("Could not open file: " + path);

    std::vector<IniSection> sections;
    IniSection current;
    bool hasCurrent = false;

    auto flush = [&]() {
        if (hasCurrent) {
            sections.push_back(std::move(current));
            current = IniSection{};
        }
    };

    std::string rawLine;
    while (std::getline(file, rawLine)) {
        // getline splits on '\n' only, so a trailing '\r' from CRLF line endings
        // needs stripping too - Trim() already eats it since '\r' is whitespace.
        std::string line = Trim(rawLine);
        if (line.empty())
            continue;

        if (line.front() == '[' && line.back() == ']') {
            flush();
            current.name = Trim(line.substr(1, line.size() - 2));
            hasCurrent = true;
            continue;
        }

        if (!hasCurrent)
            continue;

        size_t eq = line.find('=');
        if (eq == std::string::npos)
            continue;

        std::string key = Trim(line.substr(0, eq));
        std::string value = Trim(line.substr(eq + 1));
        current.entries.emplace_back(std::move(key), std::move(value));
    }
    flush();

    return sections;
}

} // namespace IniLoader
} // namespace flasc
